transformer_engine/pytorch/csrc/extensions/comm_gemm_overlap.cpp

```diff
@@ -0,0 +1,757 @@
+  NVTE_CHECK_CUDA(cudaEventRecord(_start_d2dcopy, (cudaStream_t)stream_main));
+  NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_comm, _start_d2dcopy, 0));
+  NVTE_CHECK_CUDA(cudaMemcpyAsync(ubuf_ptr, input.data_ptr(), input.numel() * input.element_size(),
+                                  cudaMemcpyDeviceToDevice, _stream_comm));
+}
+
+/*
+** Helper function to export _ubuf output.
+*/
+torch::Tensor CommOverlap::get_ubuf_output(NVTE_Comm_Overlap_Type comm_type) {
+  char *ubuf_wt_ptr = reinterpret_cast<char *>(_ubuf.data_ptr());
+  int output_c_dim0 = _ubuf.size(0);
+  if (comm_type == NVTE_Comm_Overlap_Type::REDUCE_SCATTER) {
+    ubuf_wt_ptr += (_ubuf.numel() * _ubuf.element_size() / _tp_size) * _tp_id;
+    output_c_dim0 /= _tp_size;
+  }
+  auto output_tensor =
+      torch::from_blob(ubuf_wt_ptr, {output_c_dim0, _ubuf.size(1)}, _ubuf.options());
+  return output_tensor;
+}
+
+/*
+** Helper function to set the inverse Fp8 scale for the _ubuf tensor.
+*/
+void CommOverlap::set_ubuf_scale_inv(const torch::Tensor &scale_inv) {
+  _ubuf_scale_inv_ptr = scale_inv.data_ptr();
+  _ubuf_scale_inv_initialized = true;
+}
+
+/***************************************************************************************************
+** CommOverlapP2P -- Point-2-Point (ring-exchange) comm+GEMM wrappers for PyTorch
+***************************************************************************************************/
+
+CommOverlapP2P::CommOverlapP2P(torch::Tensor sample, CommOverlapHelper &helper, int tp_size,
+                               NVTE_Comm_Overlap_Type comm_type, int num_max_streams, int cga_size,
+                               int num_comm_sms, bool set_sm_margin, bool use_ce, bool atomic_gemm,
+                               bool aggregate)
+    : te::CommOverlapP2P(helper.world_rank, helper.world_size, helper.local_rank, helper.local_size,
+                         helper.node_id, helper.num_nodes, tp_size, comm_type,
+                         std::bind(&CommOverlapHelper::allgather, helper, _1, _2, _3, _4, _5),
+                         std::bind(&CommOverlapHelper::barrier, helper, _1), num_max_streams,
+                         cga_size, num_comm_sms, set_sm_margin, use_ce, atomic_gemm, aggregate) {
+  _ubuf_bytes = sample.numel() * sample.element_size();
+  _ubuf_chunk_bytes = _ubuf_bytes / _tp_size;
+  if (_is_reduce_scatter) {
+    // GEMM + RS overlap: Allocate `2 x tp_size - 1` buffers to hold recieved GEMM chunk
+    // outputs for reduction at the end of the pipelining.
+    _ubuf_bytes = static_cast<int>((_ubuf_bytes / _tp_size) * (_tp_size * 2 - 1));
+  }
+
+  void *ubuf_ptr;
+  if (te::getenv<bool>("UB_SKIPMC")) {
+    // Multicast is disabled so we have to pre-allocate the buffer here.
+    _ubuf = torch::empty({(sample.size(0) / _tp_size) * _num_ubuf_chunks, sample.size(1)},
+                         sample.options());
+    ubuf_ptr = _ubuf.data_ptr();
+    register_gpu_buffer(&ubuf_ptr, _ubuf_bytes, false);
```

Comment:
(Just a reminder here) It seems your bugfix of legacy IPC flow is included in this PR, but the P2P part is not included.

Your bugfix: [force TE/PyTorch to always let Userbuffers manually allocate its buff…](https://github.com/NVIDIA/TransformerEngine/commit/ffb901088756ceeb52c6c96e81de9a6086b08abf)

Reply:
Thanks for catching this!